### PR TITLE
Optimizaton of block and buffer sizes in RowDiff

### PR DESCRIPTION
**Pull request by karasikov**

### CODE REVIEW

File: metagraph/src/annotation/row_diff_builder.cpp [hunk 1]
```diff
@@ -214,21 +213,34 @@ void traverse_anno_chunked(
...
+                for (uint64_t i = 0; i < block_size; ++i) {
+                    pred_chunk_idx[i + 1] = pred_chunk_idx[i];
+                    while (*pred_boundary_it == 0) {
+                        ++pred_chunk_idx[i + 1];
+                        pred_chunk.push_back(*pred_it);
+                        ++pred_it;
+                        ++pred_boundary_it;
+                    }
+                    ++pred_boundary_it;
+                }
```
Comment (hmusta) [inline metagraph/src/annotation/row_diff_builder.cpp:242]: add a comment describing what this is supposed to be doing, it's a bit hard to follow

Comment (hmusta) [reply metagraph/src/annotation/row_diff_builder.cpp:242]: Wouldn't it be easier to do this by first populating `pred_chunk_idx` by iterating through `pred_boundary_it`, then doing the copy to `pred_chunk` in a single call?

**Review by hmusta — CHANGES_REQUESTED** [review]

File: metagraph/src/annotation/row_diff_builder.cpp (hunk 1 above)
Comment (hmusta) [reply metagraph/src/annotation/row_diff_builder.cpp:242]: to make this a bit faster, does `next_bit` work with `int_vector_buffer<1>`?

Comment (hmusta) [reply metagraph/src/annotation/row_diff_builder.cpp:242]: though in the end, reading from the file will be the bottleneck

File: metagraph/src/annotation/annotation_converters.cpp [hunk 2]
```diff
@@ -1136,22 +1135,24 @@ void convert_to_row_diff(const std::vector<std::string> &files,
...
+                                            + BUF_SIZE * sizeof(uint64_t) * 2;
             if (file_size > mem_bytes) {
                 logger->warn(
                         "Not enough memory to process {}, requires {} MB, skipped",
-                        files[i], file_size/1e6);
+                        files[i], file_size / 1e6);
                 continue;
             }
-            if (file_size > mem_bytes_left || file_batch.size() >= 15'000)
+            if (file_size > mem_bytes_left)
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/annotation_converters.cpp:1155]: It doesn't keep all these files open (only num_thread at max), so there is no issue with that.

File: metagraph/src/annotation/annotation_converters.cpp [hunk 3]
```diff
@@ -1136,22 +1135,24 @@ void convert_to_row_diff(const std::vector<std::string> &files,
...
     // load as many columns as we can fit in memory, and convert them
     for (uint32_t i = 0; i < files.size(); ) {
         logger->trace("Loading columns for batch-conversion...");
         size_t mem_bytes_left = mem_bytes;
         std::vector<std::string> file_batch;
         for ( ; i < files.size(); ++i) {
-            // *2 in order to account for constructing the sparsified column
-            // TODO: *1 + size(SortedSetDisk)?
-            size_t file_size = 2 * std::filesystem::file_size(files[i]);
+            // also include a buffer for the column transformed
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/annotation_converters.cpp:1146]: added a comment

File: metagraph/src/annotation/annotation_converters.cpp [hunk 4]
```diff
@@ -1136,22 +1135,24 @@ void convert_to_row_diff(const std::vector<std::string> &files,
 
     std::filesystem::path row_reduction_fname;
 
+    const uint64_t BUF_SIZE = 500'000;
```
[the remark this answers is not in the document]
Comment (karasikov) [reply metagraph/src/annotation/annotation_converters.cpp:1138]: done

File: metagraph/src/annotation/row_diff_builder.cpp (hunk 1 above)
Comment (karasikov) [reply metagraph/src/annotation/row_diff_builder.cpp]: added comments

**Review by hmusta — APPROVED** [review]
